src/common/ult_stack_mmap.c

```diff
@@ -0,0 +1,612 @@
+	struct stack_mmap_desc *desc;
+
+	rec = (struct stack_mmap_rec *)node;
+	while ((desc = d_list_pop_entry(&rec->smr_desc_list, struct stack_mmap_desc, smd_entry)) !=
+	       NULL) {
+		int rc;
+
+		rc = munmap(desc->smd_guard_page, g_smp.smp_page_size);
+		if (unlikely(rc != 0))
+			DS_ERROR(errno,
+				 "Failed to unmap ULT stack guard page at %p: desc=%p, "
+				 "mmap_size=%zu, stack_size=%zu, page_size=%zu",
+				 desc->smd_guard_page, desc, desc->smd_rec->smr_mmap_size,
+				 desc->smd_rec->smr_stack_size, g_smp.smp_page_size);
+
+		rc = munmap(desc->smd_thread_stack, desc->smd_rec->smr_mmap_size);
+		if (unlikely(rc != 0))
+			DS_ERROR(errno,
+				 "Failed to unmap ULT stack at %p: desc=%p, mmap_size=%zu, "
+				 "stack_size=%zu",
+				 desc->smd_thread_stack, desc, desc->smd_rec->smr_mmap_size,
+				 desc->smd_rec->smr_stack_size);
+	}
+	ABT_mutex_free(&rec->smr_desc_list_mutex);
+	D_FREE(rec);
+};
+
+static int
+smd_find_insert_rec(size_t stack_size, struct stack_mmap_rec **rec)
+{
+	struct stack_mmap_rec  bt_key  = {.smr_stack_size = stack_size};
+	struct stack_mmap_rec *rec_tmp = NULL;
+	void                  *tmp;
+	int                    rc;
+
+	rc = ABT_rwlock_rdlock(g_smp.smp_rec_btree_rwlock);
+	D_ASSERT(rc == ABT_SUCCESS);
+	tmp = tfind((void *)&bt_key, &g_smp.smp_rec_btree, stack_mmap_rec_cmp);
+	rc  = ABT_rwlock_unlock(g_smp.smp_rec_btree_rwlock);
+	D_ASSERT(rc == ABT_SUCCESS);
+	if (tmp != NULL) {
+		*rec = *(struct stack_mmap_rec **)tmp;
+		D_GOTO(out, rc = -DER_SUCCESS);
+	}
+
+	D_ALLOC(rec_tmp, sizeof(struct stack_mmap_rec));
+	if (unlikely(rec_tmp == NULL))
+		D_GOTO(out, rc = -DER_NOMEM);
+
+	rec_tmp->smr_stack_size = stack_size;
+	rec_tmp->smr_mmap_size  = stack_size2mmap_size(stack_size);
+	D_INIT_LIST_HEAD(&rec_tmp->smr_desc_list);
+	rc = ABT_mutex_create(&rec_tmp->smr_desc_list_mutex);
+	if (unlikely(rc != ABT_SUCCESS)) {
+		D_ERROR("Failed to create ABT mutex: " AF_RC "\n", AP_RC(rc));
+		D_GOTO(error_rec_tmp, rc = dss_abterr2der(rc));
+	}
+
+	rc = ABT_rwlock_wrlock(g_smp.smp_rec_btree_rwlock);
+	D_ASSERT(rc == ABT_SUCCESS);
+	tmp = (struct stack_mmap_rec *)tsearch(rec_tmp, &g_smp.smp_rec_btree, stack_mmap_rec_cmp);
+	rc  = ABT_rwlock_unlock(g_smp.smp_rec_btree_rwlock);
+	D_ASSERT(rc == ABT_SUCCESS);
+
+	if (unlikely(tmp == NULL)) {
+		DL_ERROR(-DER_NOMEM, "Failed to create new btree node");
+		D_GOTO(error_mutex, rc = -DER_NOMEM);
+	}
+
+	if (unlikely(*(struct stack_mmap_rec **)tmp != rec_tmp)) {
+		ABT_mutex_free(&rec_tmp->smr_desc_list_mutex);
+		D_FREE(rec_tmp);
+		rec_tmp = *(struct stack_mmap_rec **)tmp;
+	}
+
+	*rec = rec_tmp;
+	D_DEBUG(DB_MEM, "New btree record %p of size %zu (mmap size %zu)\n", *rec,
+		(*rec)->smr_stack_size, (*rec)->smr_mmap_size);
+	D_GOTO(out, rc = -DER_SUCCESS);
+
+error_mutex:
+	ABT_mutex_free(&rec_tmp->smr_desc_list_mutex);
+error_rec_tmp:
+	D_FREE(rec_tmp);
+out:
+	return rc;
+}
+
+static int
+smd_find_insert_desc(size_t stack_size, struct stack_mmap_desc **desc)
```

Comment:
Can we use a different prefix for those functions? We've already have a SMD component (see src/include/daos_srv/smd.h). :)

Reply:
- [ ] Change `smd_` prefix